text/0002-group-spotify-playlists.md

```diff
@@ -0,0 +1,204 @@
+
+Finally, we want to take the features prescribed by the desired theme and the features that are
+most common among users, and pull songs that match as many of these features as possible.
+
+Playlist generation would of course be performed on the server side. Also, the playlist would be
+initially created under a Spotify account dedicated to our app and subsequently shared with every
+member of the lobby. This way, we don't have to synchronize multiple users' libraries, and updates
+would only apply to a single playlist. This makes our lives a lot easier.
+
+In order to perform playlist modification, we will first authenticate our server and obtain a 
+Spotify Refresh Token. From there, we can use the Refresh Token to generate a new Access Token
+to perform requests onto the Spotify API. We **should** batch all of our operations to reduce
+the amount refreshes required.
+
+[Spotify API]: https://developer.spotify.com/documentation/web-api/
+
+#### Playing music from the playlist
+
+Spotify offers a [playback SDK] to play music from Spotify on the client side. Combined with the
+[web API] for playback controls, we have a way for music to be played directly from our application.
+
+[playback SDK]: https://developer.spotify.com/documentation/web-playback-sdk/
+[web API]: https://developer.spotify.com/documentation/web-api/
+
+It is important to note that the playback SDK is only functional for premium users. As such, when a
+user logs in, if they are determined to not have a Spotify Premium account then we would need to
+inform them that playback from our app is not allowed.
+
+### What are the requirements for this project?
+
+The PWA itself will probably be built in Next.js (strictly speaking, this won't be necessary for
+our MVP (see below), but it will be a good idea for future-proofing), and the backend will be
+Node.js-based. If we proceed with using Firestore for handling user lobbies to start with, then anyone with Firebase experience would be handy.
+
+We have two options for generating playlists:
+* We could use OCI to host a server for generating playlists.
+* We could use a serverless system.
+
+If we use a server-based model using something like OCI, then we would need to set up more
+infrastructure. This might be a good learning experience for those members who are less experienced
+with OCI, and it also may be cool to show the OCI team what we've been working on so we can get
+more free stuff. Also, OCI makes it easier to scale.
+
+If we use serverless, then we would need to set up a way to prevent too many playlist updates at
+once. One way to accomplish that is:
+1. Whenever a user logs in, save the user's data that is needed for playlist creation into their
+   appropriate entry in Firestore.
+2. Run a pub/sub serverless function that checks for updates every hour in a given user's data
+   field. If the data is constant, we can skip the update; otherwise, we run the update to the
+   playlist.
```

Comment:
Sounds good to me, just wanna clarify, these updates for a given user's data are only ever gonna happen when the user logs in right, since we're not storing refresh tokens or anything?